### PR TITLE
PODS-4738 Create penalties tile on PSA dashboard

### PR DESCRIPTION
**Pull request by grantrobson**

### CODE REVIEW

File: app/controllers/financialStatement/PenaltiesPartialController.scala [hunk 1]
```diff
@@ -47,9 +46,16 @@ class PenaltiesPartialController @Inject()(
   def penaltiesPartial: Action[AnyContent] = identify.async { implicit request =>
 
     fsConnector.getPsaFS(request.psaIdOrException.id).flatMap { psaFS =>
-      val json = Json.obj("displayLink" -> Json.toJson(psaFS.nonEmpty),
-      "viewPenaltiesUrl" -> config.viewPenaltiesUrl)
-      renderer.render(template = "partials/penalties.njk", json).map(Ok(_))
+      val result = if (psaFS.isEmpty) {
+        Future.successful(play.twirl.api.Html(""))
```
Comment (vinnicombe) [inline app/controllers/financialStatement/PenaltiesPartialController.scala:50]: Import

Comment (grantrobson) [reply app/controllers/financialStatement/PenaltiesPartialController.scala]: done

File: app/services/paymentsAndCharges/PaymentsAndChargesService.scala [hunk 2]
```diff
@@ -98,10 +98,12 @@ class PaymentsAndChargesService @Inject()() {
...
-  def getUpcomingCharges(schemeFS: Seq[SchemeFS]): Seq[SchemeFS] =
-    schemeFS
-      .filter(_.dueDate.nonEmpty)
-      .filter(_.dueDate.get.isAfter(DateHelper.today))
+  def extractUpcomingCharges[A](seqOfCharges: Seq[A], dateToFilterOn: A => Option[LocalDate]): Seq[A] =
+    seqOfCharges.filter{ charge =>
+      val optionDate = dateToFilterOn(charge)
+      optionDate.nonEmpty &&
+        (optionDate.get.isEqual(DateHelper.today) || optionDate.get.isAfter(DateHelper.today))
+    }
```
Comment (vinnicombe) [inline app/services/paymentsAndCharges/PaymentsAndChargesService.scala:106]: Did this function need to be completely refactored and renamed? Does it need to be generic? Could just adding `.filter(_.dueDate.get.isEqual(DateHelper.today))` have done the job?

Comment (grantrobson) [reply app/services/paymentsAndCharges/PaymentsAndChargesService.scala:106]: It's used by both payments and penalties so I couldn't see a way other than duplicating the method to allow both of these to use the same method.

File: test/controllers/financialStatement/PenaltiesPartialControllerSpec.scala [hunk 3]
```diff
@@ -44,23 +47,49 @@ class PenaltiesPartialControllerSpec extends ControllerSpecBase with NunjucksSup
...
+        ),
+        Json.obj(
+          "total" -> "£2,058.10",
+          "span" -> "Total overdue payments:"
+        )
+      )
+
+    val links = Seq(
+      Link(id = "aft-penalties-id",
+        url = controllers.financialStatement.routes.SelectSchemeController.onPageLoad(year = "2020").url,//"http://localhost:8206/manage-pension-scheme-accounting-for-tax/2020/select-a-scheme",
```
Comment (vinnicombe) [inline test/controllers/financialStatement/PenaltiesPartialControllerSpec.scala:78]: Comment

Comment (grantrobson) [reply test/controllers/financialStatement/PenaltiesPartialControllerSpec.scala]: removed

File: app/services/paymentsAndCharges/PaymentsAndChargesService.scala (hunk 2 above)
Comment (vinnicombe) [reply app/services/paymentsAndCharges/PaymentsAndChargesService.scala:106]: Ah yes, I see it being used for PsaFS. So it does need to be generic. Is rest of the refactor necessary?

Comment (vinnicombe) [reply app/services/paymentsAndCharges/PaymentsAndChargesService.scala:106]: Actually I see without a defined type at the point of filtering, you won't be able to filter as it is currently

Comment (grantrobson) [reply app/services/paymentsAndCharges/PaymentsAndChargesService.scala:106]: Well that's a moot point. I did consider that when I was doing it. We could just duplicate the method, i.e. one for schemes one for penalties, but then we have code duplication.

**Review by vinnicombe — APPROVED** [review]

**Pull request merged**